webhook_service/main.go

```diff
@@ -0,0 +1,873 @@
+func SubscribeTransients(rdb *redis.Client, mu *sync.Mutex) {
+	ctx := context.Background()
+	pubsub := rdb.Subscribe(ctx, transientChannel)
+	defer pubsub.Close()
+
+	logger.Info(fmt.Sprintf("Subscribing to: %s", transientChannel))
+	for msg := range pubsub.Channel() {
+		var transientMsg Message
+		if err := json.Unmarshal([]byte(msg.Payload), &transientMsg); err != nil {
+			logger.Error(fmt.Sprintf("transient unmarshal error: %v, JSON: %s", err, msg.Payload))
+			continue
+		}
+		go ProcessTransient(rdb, transientMsg, mu)
+	}
+}
+
+func main() {
+	logger.Info("Starting service")
+	rdb := SetupRedis()
+
+	// For the Docker health check
+	path := filepath.Join("/app", "webhook-service-healthy")
+	logger.Info(path)
+	exec.Command("touch", path).Start()
+
+	// A context is required to work with redis
+	ctx := context.Background()
+	var mu sync.Mutex
+
+	// Monitor the input queue and dispatch messages to the appropriate queue worker
+	go func() {
+		for {
+			results, err := rdb.BLPop(ctx, 0, inputQueue).Result()
+			if err != nil {
+				logger.Error(fmt.Sprintf("Input pop error: %v", err))
+				time.Sleep(time.Second)
+				continue
+			}
+
+			var msg Message
+			var rawStr string
+			rawMessage := []byte(results[1])
+			// This is stupid but we have to get the raw JSON string into a string before we can
+			// unmarshal it into our struct because of how it's encoded from CRCON
+			if err := json.Unmarshal(rawMessage, &rawStr); err == nil {
+				if err := json.Unmarshal([]byte(rawStr), &msg); err != nil {
+					logger.Error(fmt.Sprintf("unmarshal error: %v, JSON: %s", err, rawStr))
+					time.Sleep(time.Second)
+					continue
+				}
+			} else {
+				logger.Error(fmt.Sprintf("Input unmarshal error: %v, JSON: %s", err, rawMessage))
+				time.Sleep(time.Second)
+				continue
+			}
+
+			// Check to see if we already know this webhook's rate limit bucket
+			// We check by webhook ID since multiple webhooks can share the same bucket
+			webhookID, err := ExtractWebhookID(msg.Payload.URL)
+			if err != nil {
+				logger.Error(fmt.Sprintf("Could not parse webhook ID from: %s:%s", msg.Payload.URL, err))
+				continue
+			}
+			bucket, err := rdb.HGet(ctx, bucketsHash, webhookID).Result()
+			var msgBytes []byte
+			msgBytes, _ = json.Marshal(msg)
+			if err == redis.Nil {
+				// Unknown webhook, send to first-time queue
+				rdb.RPush(ctx, firstTimeQueue, msgBytes)
+				rdb.LTrim(ctx, firstTimeQueue, 0, maxQueueSize-1)
+			} else {
+				// Known bucket, send to bucket queue
+				queueKey := bucketQueuePrefix + bucket
+				rdb.RPush(ctx, queueKey, msgBytes)
+				rdb.LTrim(ctx, queueKey, 0, maxQueueSize-1)
+
+				// Start workers if new bucket
+				if _, exists := GetWorker(bucket, &mu); !exists {
+					worker := NewBucketWorker(bucket, rdb)
+					SetWorker(bucket, worker, &mu)
+					go worker.ProcessQueue(rdb)
+				}
+			}
+		}
+	}()
+
+	go ProcessFirstTime(rdb, &mu)
+	go SubscribeTransients(rdb, &mu)
+
+	// Keep running
+	select {}
```

Comment:
whenever there is a SIGTERM or SIGINT, the program should gracefully shutdown.Closing all redis connections, cleaning up any other stuff if necessary.

E.g. like here:
https://github.com/FlorianSW/hll-discord-server-watcher/blob/main/cmd/cmd.go#L94-L101
(just with closing the redis client and stuff)